overlord/configstate/configcore/cloud.go

```diff
@@ -0,0 +1,90 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package configcore
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"os"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/overlord/auth"
+	"github.com/snapcore/snapd/overlord/state"
+)
+
+func alreadySeeded(tr Conf) (bool, error) {
+	st := tr.State()
+	st.Lock()
+	defer st.Unlock()
+	var seeded bool
+	err := tr.State().Get("seeded", &seeded)
+	if err != nil && err != state.ErrNoState {
+		return false, err
+	}
+	return seeded, nil
+}
+
+func handleCloud(tr Conf) error {
+	// if we are during seeding try to capture cloud information
+	seeded, err := alreadySeeded(tr)
+	if err != nil {
+		return err
+	}
+	if seeded {
+		// already done
+		return nil
+	}
+
+	data, err := ioutil.ReadFile(dirs.CloudInstanceDataFile)
+	if os.IsNotExist(err) {
+		// nothing to do
+		return nil
+	}
+	if err != nil {
+		logger.Noticef("cannot read cloud instance information %q: %v", dirs.CloudInstanceDataFile, err)
+		return nil
+	}
+	var instanceData struct {
```

Comment:
cloud-init will support this json content in /run/cloud-init/instance-data.json as sourced. We may add more fields under the generalized 'v1'  key, but we will not change any existing key names under 'v1'. This approach is safe and supported.